packages/api/src/utils/highlightGenerator.ts

```diff
@@ -0,0 +1,328 @@
+import { diff_match_patch as DiffMatchPatch } from 'diff-match-patch'
+import { interpolationSearch } from './interpolationSearch'
+import { v4 as uuidv4 } from 'uuid'
+import { nanoid } from 'nanoid'
+
+const highlightTag = 'omnivore_highlight'
+export const maxHighlightLength = 2000
+
+const nonParagraphTagsRegEx =
+  /^(a|b|basefont|bdo|big|em|font|i|s|small|span|strike|strong|su[bp]|tt|u|code|mark)$/i
+const highlightContentRegex = new RegExp(
+  `<${highlightTag}>([\\s\\S]*)<\\/${highlightTag}>`,
+  'i'
+)
+const maxDeepPatchDistance = 4000
+const maxDeepPatchThreshhold = 0.5
+const maxSurroundingTextLength = 2000
+
+type TextNode = {
+  startIndex: number
+  node: Node
+  isParagraphStart?: boolean
+}
+
+type ArticleTextContent = {
+  textNodes: TextNode[]
+  articleText: string
+}
+
+export type EmbeddedHighlightData = {
+  prefix: string
+  suffix: string
+  quote: string
+  id: string
+  shortId: string
+  patch: string
+}
+
+function getTextNodesBetween(rootNode: Node, startNode: Node, endNode: Node) {
+  let textNodeStartingPoint = 0
+  let articleText = ''
+  let newParagraph = false
+  const textNodes: TextNode[] = []
+  let pastStartNode = false,
+    reachedEndNode = false
+
+  function pushNode(node: Node) {
+    textNodes.push({
+      node,
+      startIndex: textNodeStartingPoint,
+      isParagraphStart: newParagraph,
+    })
+    textNodeStartingPoint += node.nodeValue?.length || 0
+    articleText += node.nodeValue
+    newParagraph = false
+  }
+
+  function getTextNodes(node: Node) {
+    if (node == startNode) {
+      pastStartNode = true
+    }
+
+    if (node.nodeType == 3) {
+      if (
+        pastStartNode &&
+        !reachedEndNode &&
+        !/^\s*$/.test(node.nodeValue || '')
+      ) {
+        pushNode(node)
+      }
+    } else {
+      if (!nonParagraphTagsRegEx.test((node as Element).tagName))
+        newParagraph = true
+    }
+
+    for (
+      let i = 0, len = node.childNodes.length;
+      !reachedEndNode && i < len;
```

Comment:
I guess we could check `reachedEndNode` before checking the `nodeType` on line 63

Reply:
Let me leave this one as-is for now, I think i need to add some extra tests for the paragraph wrapping stuff, and it could effect the else block below.